pkg/blob/controllerserver.go

```diff
@@ -768,18 +774,75 @@ func (d *Driver) copyBlobContainer(_ context.Context, req *csi.CreateVolumeReque
 	vs := req.VolumeContentSource
 	switch vs.Type.(type) {
 	case *csi.VolumeContentSource_Snapshot:
 		return status.Errorf(codes.InvalidArgument, "copy volume from volumeSnapshot is not supported")
 	case *csi.VolumeContentSource_Volume:
-		return d.copyBlobContainer(ctx, req, accountKey, dstContainerName, storageEndpointSuffix)
+		return d.copyBlobContainer(ctx, req, accountSASToken, dstContainerName, storageEndpointSuffix)
 	default:
 		return status.Errorf(codes.InvalidArgument, "%v is not a proper volume source", vs)
 	}
 }
 
+func (d *Driver) authorizeAzcopyWithIdentity() error {
+	azureAuthConfig := d.cloud.Config.AzureAuthConfig
+	if len(azureAuthConfig.AADClientSecret) > 0 {
+		klog.V(2).Infof("use service principal to authorize azcopy")
+		if err := os.Setenv(azcopyAutoLoginType, "SPN"); err != nil {
+			return err
+		}
+		if azureAuthConfig.AADClientID == "" {
+			return fmt.Errorf("AADClientID and AADClientSecret must be set when use service principal")
+		}
+		if err := os.Setenv(azcopySPAApplicationID, azureAuthConfig.AADClientID); err != nil {
+			return err
+		}
+		if err := os.Setenv(azcopySPAClientSecret, azureAuthConfig.AADClientSecret); err != nil {
+			return err
+		}
+		if azureAuthConfig.TenantID != "" {
+			if err := os.Setenv(azcopyTenantID, azureAuthConfig.TenantID); err != nil {
+				return err
+			}
+			klog.V(2).Infof(fmt.Sprintf("set AZCOPY_TENANT_ID=%s successfully", azureAuthConfig.TenantID))
+		}
+		return nil
+	}
+	if azureAuthConfig.UseManagedIdentityExtension {
+		if err := os.Setenv(azcopyAutoLoginType, "MSI"); err != nil {
+			return err
+		}
+		if len(azureAuthConfig.UserAssignedIdentityID) > 0 {
+			klog.V(2).Infof("use user assigned managed identity to authorize azcopy")
+			err := os.Setenv(azcopyMSIClientID, azureAuthConfig.UserAssignedIdentityID)
+			return err
+		}
+		klog.V(2).Infof("use system-assigned managed identity to authorize azcopy")
+		return nil
+	}
+	return fmt.Errorf("AADClientSecret shouldn't be nil or useManagedIdentityExtension must be set to true")
+}
+
+func (d *Driver) getSASToken(ctx context.Context, accountName, accountKey, storageEndpointSuffix string, accountOptions *azure.AccountOptions, secrets map[string]string, secretName, secretNamespace string) (string, error) {
+	if len(secrets) > 0 || len(d.cloud.Config.AzureAuthConfig.AADClientSecret) == 0 && !d.cloud.Config.AzureAuthConfig.UseManagedIdentityExtension {
+		var err error
+		if accountKey == "" {
+			if _, accountKey, err = d.GetStorageAccesskey(ctx, accountOptions, secrets, secretName, secretNamespace); err != nil {
+				return "", status.Errorf(codes.Internal, "failed to GetStorageAccesskey on account(%s) rg(%s), error: %v", accountOptions.Name, accountOptions.ResourceGroup, err)
+			}
+		}
+		klog.V(2).Infof("generate sas token for account(%s)", accountName)
+		var accountSASToken string
```

Comment:
`return generateSASToken(accountName, accountKey, storageEndpointSuffix, d.sasTokenExpirationMinutes)`, that's more clear

Reply:
done